【Paddle Tensor No.8、9、14、15】为Tensor新增`__rshift__`,`__lshift__`,`__rlshift__`,`__rrshift__`

MrXnneHang opened the pull request:
### PR Category
User Experience

### PR Types
New features


### Description

我是来提我发现的问题的：

[第二次正经PR : 发现左右操作数类型的冲突Bug和可选解决方式 ,`__rrshift__`和`__rshift__`,`__rlshift__`和`__lshift__`.](https://xnnehang.top/blog/163)

@HydrogenSulfate 

问题复现：

```python
>>> import paddle
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
>>> data = paddle.to_tensor([2,4,8])
>>> shift = paddle.to_tensor([1])
>>> data << shift
Tensor(shape=[3], dtype=int64, place=Place(cpu), stop_gradient=True,
       [4 , 8 , 16])
>>> shift << data
Tensor(shape=[3], dtype=int64, place=Place(cpu), stop_gradient=True,
       [4  , 16 , 256])
>>> 1 << data
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/paddle/tensor/math.py", line 7879, in right_left_shift
    return bitwise_left_shift(x, y, is_arithmetic, out, name)
  File "/usr/local/lib/python3.9/dist-packages/paddle/tensor/math.py", line 7651, in bitwise_left_shift
    return _C_ops.bitwise_left_shift(x, y, is_arithmetic)
ValueError: (InvalidArgument) bitwise_left_shift(): argument 'y' (position 1) must be Tensor, but got int (at /paddle/paddle/fluid/pybind/eager_utils.cc:1330)
```

单纯复用会有这种情况.`1<<data`的时候应该调用data的`__rlshift__`，但是一调用就报错。如果支持int的话就不会有这个问题，参见我上面的链接🔗。

torch是支持的:

```Python
>>> import torch
>>> data = torch.Tensor([2,4,8]).to(torch.int)
>>> shift = torch.Tensor([1]).to(torch.int)
>>> data << shift
tensor([ 4,  8, 16], dtype=torch.int32)
>>> data << 1
tensor([ 4,  8, 16], dtype=torch.int32)
```

- task: #69082

MrXnneHang committed:
update

MrXnneHang commented:
更新了链接

HydrogenSulfate commented:
![image](https://github.com/user-attachments/assets/01e308dd-14e7-4d6b-bda3-1921729584e4)
这个示例本身不能说明torch能支持python int吧？两个operand都是torch.Tensor，至少要加一个 `1 << torch.Tensor`

MrXnneHang commented:
![image](https://github.com/user-attachments/assets/2b698e96-f7dc-490f-9455-4c0e565759b8)

说起来它把左右操作数都自动转换成Tensor了

![image](https://github.com/user-attachments/assets/e74adc73-da7b-4eeb-ae49-3319b45cb296)

numpy的逻辑也一致。

所以我觉得应该得有一个AutoTensor的方法，把int尝试转成Tensor(dtype = paddle.int,shape=[1])

MrXnneHang commented:
另外暂时也只需要考虑int->Tensor的情况即可,torch和numpy也都没有对float对额外的类型转换

```python
>>> import torch
>>> data = torch.Tensor([2,4,8])
>>> data << 1.0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "lshift_cpu" not implemented for 'Float'
```

```python
>>> import numpy as np
>>> data = np.array([2,4,8])
>>> data << 1.0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'left_shift' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

float没有被转换成Tensor

MrXnneHang commented:
另外还有这种情况，如果发生移位失败就会相应位置变成True/False

```python
>>> import numpy
>>> data = numpy.array([2,4,8])
>>> 1 > data
array([False, False, False])
```

```python
>>> import numpy as np
>>> data = np.array([1,2,4])
>>> 2 > data
array([ True, False, False])
>>> 4 > data
array([ True,  True, False]) 
```

MrXnneHang committed:
fix:update expected output

HydrogenSulfate commented:
这个PR是不是就是这几个API？
![image](https://github.com/user-attachments/assets/62f2fdb4-25e5-43e5-991b-71ea7580da00)

MrXnneHang commented:
包括复用bitwise_left_shift,biwise_right_shift的__rshift__和__lshift__

HydrogenSulfate commented:
> 包括复用bitwise_left_shift,biwise_right_shift的__rshift__和__lshift__

好的

HydrogenSulfate commented:
> 另外暂时也只需要考虑int->Tensor的情况即可,torch和numpy也都没有对float对额外的类型转换
> 
> ```python
> >>> import torch
> >>> data = torch.Tensor([2,4,8])
> >>> data << 1.0
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: "lshift_cpu" not implemented for 'Float'
> ```
> 
> ```python
> >>> import numpy as np
> >>> data = np.array([2,4,8])
> >>> data << 1.0
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> TypeError: ufunc 'left_shift' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
> ```
> 
> float没有被转换成Tensor

对的，移位操作一般只用于int，float是不需要的

HydrogenSulfate commented on python/paddle/tensor/__init__.py:
```diff
@@ -862,4 +866,8 @@
+    ('__lshift__', 'left_shift'),
+    ('__rshift__', 'right_shift'),
+    ('__rlshift__', 'right_left_shift'),
+    ('_rrshift__', 'right_right_shift'),
```
`_rrshift__`这里是不是少了个下划线？

MrXnneHang replied:
对。

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,131 @@ def bitwise_right_shift_(
+        Tensor: Result of ``__rshift__`` . It is a N-D Tensor with the same data type of input Tensor.
+    Examples:
+        .. code-block:: python
+            :name: rshift_example1
```
这里的`.. xxx` 代码块之间是不是应该有一个空行？

MrXnneHang replied:
好

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,131 @@ def bitwise_right_shift_(
+                        [8 , 8 , 16, 32]])
+    """
+    # Directly call bitwise_left_shift
+    return bitwise_left_shift(x, y, is_arithmetic, out, name)
```
1. `left_shift`不是API标准里面的接口，不能露给用户
2. 这里是纯转发，是否能直接使用`bitwise_left_shift`，删掉`left_shift`

MrXnneHang replied:
这里我有想法。
我正在打算定义这样一个op:
```yaml
- op: auto_tenosor
  args: (int x)
  output: Tensor(out)
  infer_meta:
    func: AutoTensorMeta
  kernel:
    func: auto_tensor
  traits: paddle::dialect::ForwardOnlyTrait
```
我需要组合复用bitwise_left_shift和auto_tensor来实现对int的自动转换，所以我需要保留left_shift.

MrXnneHang replied:
然后left_shift可以重载为(Tensor,int),(int,Tenosr)和(Tensor,Tensor)

要下动车了，暂时可能回复不了

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,131 @@ def bitwise_right_shift_(
+    # equals to: 1 >> y
+    # expected output:Tensor([0,0,0])
+    # But it will raise TypeError,doesn't support for `int`
+    return bitwise_right_shift(x, y, is_arithmetic, out, name)
```
三个API同理

MrXnneHang commented:
这里只是我当时复现问题时候的代码，docstring我最后都会修改和重写。

目前我正在尝试定义一个将int自动转换为Tensor的kernel:

```
- op: auto_tenosor
  args: (int x)
  output: Tensor(out)
  infer_meta:
    func: AutoTensorMeta
  kernel:
    func: auto_tensor
  traits: paddle::dialect::ForwardOnlyTrait
```
命名可以再改

然后我使用它和`_C_ops.bitwise_left_shift`和`_C_ops.bitwise_right_shift`来实现重载了(int,Tensor),(Tensor,int),(Tensor,Tensor)的left_shift和right_shift然后用来实现`__lshift__`和`__rshift__`,以及`__rlshift__`和`__rrshift__`。我这个重载是在python上面做的，这样可以避免动到`bitwise_left_shift`的原生代码。

最后我希望实现的功能是像这样:

```python
>>> import paddle
>>> data = paddle.to_tensor([2,4,8],dtype=paddle.int32)
>>> shift = torch.Tensor([1],dtype=paddle.int32)
>>> data << shift
tensor([ 4,  8, 16], dtype=paddle.int32)
>>> data << 1
tensor([ 4,  8, 16], dtype=paddle.int32)
>>> 1 << data 
tensor([4,16,256],dtype=paddle.int32)
>>> shift << data
tensor([4,16,256],dtype=paddle.int32)
```
目前的直接复用是不支持`data << 1` , ` 1 << data`的写法的。

等我觉得写好了`@`您来review.到时候我们再讨论一下testing和API标准的问题。

MrXnneHang committed:
fix:`_rrshift__` -> `__rrshift__`

HydrogenSulfate commented:
> 这里只是我当时复现问题时候的代码，docstring我最后都会修改和重写。
> 
> 目前我正在尝试定义一个将int自动转换为Tensor的kernel:
> 
> ```
> - op: auto_tenosor
>   args: (int x)
>   output: Tensor(out)
>   infer_meta:
>     func: AutoTensorMeta
>   kernel:
>     func: auto_tensor
>   traits: paddle::dialect::ForwardOnlyTrait
> ```
> 
> 命名可以再改
> 
> 然后我使用它和`_C_ops.bitwise_left_shift`和`_C_ops.bitwise_right_shift`来实现重载了(int,Tensor),(Tensor,int),(Tensor,Tensor)的left_shift和right_shift然后用来实现`__lshift__`和`__rshift__`,以及`__rlshift__`和`__rrshift__`。我这个重载是在python上面做的，这样可以避免动到`bitwise_left_shift`的原生代码。
> 
> 最后我希望实现的功能是像这样:
> 
> ```python
> >>> import paddle
> >>> data = paddle.to_tensor([2,4,8],dtype=paddle.int32)
> >>> shift = torch.Tensor([1],dtype=paddle.int32)
> >>> data << shift
> tensor([ 4,  8, 16], dtype=paddle.int32)
> >>> data << 1
> tensor([ 4,  8, 16], dtype=paddle.int32)
> >>> 1 << data 
> tensor([4,16,256],dtype=paddle.int32)
> >>> shift << data
> tensor([4,16,256],dtype=paddle.int32)
> ```
> 
> 目前的直接复用是不支持`data << 1` , ` 1 << data`的写法的。
> 
> 等我觉得写好了`@`您来review.到时候我们再讨论一下testing和API标准的问题。

应该不需要这么麻烦，一个简单的左移例子(demo，没有添加必要的检查，且没有按照规范写的代码，只是尝试触发右运算数的 `__rlshift__`)：

``` py
import paddle



def __rlshift__(self: paddle.Tensor, other: int | paddle.Tensor):
    return  paddle.bitwise_left_shift(
        paddle.to_tensor(other),
        self,
    )


setattr(paddle.Tensor, '__rlshift__', __rlshift__)

two = paddle.to_tensor(2)

print(10 << two)
```

MrXnneHang commented:
好的，我尝试一下。

HydrogenSulfate commented:
> ardOnlyTrait

调用`rop`系列运算符的原理看起来是会先调用左操作数的`op`，如果`op`抛出`TypeError`异常，则会被python捕获，此时会开始尝试调用右操作数的`rop`，所以应该只需要在撰写的 `rop` 函数中对拿到的左操作数类型进行判断（我上面代码里类型提示写错了，左操作数不会是`Tensor`类型，如果是`Tensor`肯定调的是`op`而不会轮到`rop`）
1. 如果是float，则抛出TypeError，与python的报错保持一致
2. 如果是int，则调用to_tensor转为0-D Tensor，然后调用bitwise op进行运算，bitwise op会在内部完成隐式广播，最后返回结果

MrXnneHang committed:
feat: use `paddle.to_tensor` to support int input,and raise TypeError for float

MrXnneHang commented:
```python
>>> import paddle
>>> data = paddle.to_tensor([1,2,4],dtype=paddle.int32)
>>> shift = paddle.to_tensor(1,dtype=paddle.int32)
>>> data << shift
Tensor(shape=[3], dtype=int32, place=Place(cpu), stop_gradient=True,
       [2, 4, 8])
>>> data << 1
Tensor(shape=[3], dtype=int32, place=Place(cpu), stop_gradient=True,
       [2, 4, 8])
>>> shift << data
Tensor(shape=[3], dtype=int32, place=Place(cpu), stop_gradient=True,
       [2 , 4 , 16])
>>> 1 << data
Tensor(shape=[3], dtype=int32, place=Place(cpu), stop_gradient=True,
       [2 , 4 , 16])
>>> data << 1.0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/paddle/tensor/math.py", line 7826, in __lshift__
    raise TypeError("unsupported operand type(s) for <<:'float'")
TypeError: unsupported operand type(s) for <<:'float'
>>> 1.0 << data
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/paddle/tensor/math.py", line 7934, in __rlshift__
    raise TypeError("unsupported operand type(s) for <<:'float'")
TypeError: unsupported operand type(s) for <<:'float'
>>> 
```

现在基本上和torch以及numpy一致了。

我不是很清楚是不是可以直接在全局定义域里而不是类里定义魔术方法像`__rshift__`，以及关于API标准的问题，如果需要修改请告诉我。

我现在正要添加testing,我是加到`test_bitwise_shift_op`中，还是应该新建一个`test_tensor_shift_op`,然后模仿`test_bitwise_shift_op`来一个全面的testing。

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+            Tensor(shape=[2, 4], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[20  , 80  , 320 , 1280],
+                    [64  , 136 , 128 , 130 ]])
+
+
```
```suggestion

```

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+            Tensor(shape=[2, 4], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[20 , 40 , 80 , 160],
+                    [32 , 34 , 64 , 130]])
+
```
```suggestion
```

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+            >>> import paddle
+            >>> x = paddle.to_tensor([[1,2,3,4,], [2,3,2,1]])
+            >>> 8 >> x  # it will call x.__rrshift__(64)
+            Tensor(shape=[2, 4], dtype=int64, place=Place(cpu), stop_gradient=True,
+                    [[4, 2, 1, 0],
+                        [2, 1, 2, 4]])
```
这里标准输出看起来换行没对齐，可以确认下，其他地方同理

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+
+            >>> import paddle
+            >>> x = paddle.to_tensor([[1,2,3,4,], [2,3,2,1]])
+            >>> 8 >> x  # it will call x.__rrshift__(64)
```
rrshift(64)?不对吧，是rrshift(8)吧

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+    if isinstance(y, int):
+        y = paddle.to_tensor(y, dtype=x.dtype)
+    elif isinstance(y, float):
+        raise TypeError("unsupported operand type(s) for >>:'float'")
```
```suggestion
        raise TypeError("unsupported operand type(s) for >>: 'float' and 'Tensor'")
```
其他地方同理，修改一下

HydrogenSulfate reviewed (commented):
再添加一些单测即可

MrXnneHang commented:
好的

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,220 @@ def bitwise_right_shift_(
+    out: Tensor | None = None,
+    name: str | None = None,
+) -> Tensor:
+    r"""
```
魔术方法不是公开的API，不用docstring，其他函数的docsting也删除

HydrogenSulfate commented:
单测左操作数使用包含正负零的整数随机数，并且需要覆盖int64和int32的范围测试（如 `1 << tensor(34, int32)` 和 `1 << tensor(62, int64)`）

MrXnneHang committed:
:memo: docs:remove docstring of magic method

MrXnneHang committed:
test:add test for `__lshift__`,`__rshift__`,`__rlshift__`,`__rrshift__`,`__rshift__`cannot pass ref_right_shift_logical

`__rshift__` cannot pass TestBitwiseRightShiftAPI.test_static_api_logical and test_dygraph_api_arithmetic

SigureMo commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,63 @@ def bitwise_right_shift_(
+def __lshift__(
+    x: Tensor,
+    y: Tensor | int,
+    is_arithmetic: bool = True,
+    out: Tensor | None = None,
+    name: str | None = None,
```
`__lshift__` 签名有这么复杂么？后三个参数应该没有吧？

https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__lshift__.html#lshift

Array API 规范也没有

MrXnneHang committed:
test: use `x.__rrshift__(y,False)` to send is_arithmetic

MrXnneHang committed:
test: keep it as the same.

使用__rshift__而不是__rrshift__,使用__lshift__,而不是__rlshift__

MrXnneHang committed:
chore: remove `out` and `name`

MrXnneHang committed:
fix: 直接继承TestCase来进行不同类型测试

MrXnneHang committed:
test: add more test in uint8,int8,int16 for `__rlshift__` and `__rrshift__`

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,55 @@ def bitwise_right_shift_(
+    if isinstance(y, int):
+        y = paddle.to_tensor(y, dtype=x.dtype)
+    elif isinstance(y, float):
+        raise TypeError("unsupported operand type(s) for <<:'float'")
```
注意下异常信息规范，这些异常信息的打印建议也放到单测中捕获并断言是否正确。
![image](https://github.com/user-attachments/assets/31e651d8-a2e5-4174-bd40-19cb08d424ce)

```suggestion
        raise TypeError("unsupported operand type(s) for <<: 'float' and 'Tensor'")
```

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,55 @@ def bitwise_right_shift_(
+    if isinstance(y, int):
+        y = paddle.to_tensor(y, dtype=x.dtype)
+    elif isinstance(y, float):
+        raise TypeError("unsupported operand type(s) for >>:'float'")
```
同上

HydrogenSulfate commented on python/paddle/tensor/math.py:
```diff
@@ -7769,6 +7769,55 @@ def bitwise_right_shift_(
+    if isinstance(y, int):
+        y = paddle.to_tensor(y, dtype=x.dtype)
+    elif isinstance(y, float):
+        raise TypeError("unsupported operand type(s) for <<:'float'")
```
同上

MrXnneHang committed:
typos: raises TypeError between `'float' and 'Tensor'` and `'Tensor' and 'float'`

MrXnneHang committed:
test: add test for `TypeError` in float

MrXnneHang committed:
typo: `bool` -> `float`

HydrogenSulfate reviewed (approved):
LGTM
